### PR TITLE
Testing of the Marine DA jjobs

### LINKED ISSUE #232: 5deg Marine 3DVar ctest using j-job
Same as #229 but for ocean and sea-ice.

**Remaining issues, notes to self:**
- [x] yaml file passed to exp setup py script is still hard-coded ... do a sed or something
~- [ ] Fixing this: _line 41: setpdy.sh: command not found_ would make the test shell script a bit cleaner~
- [x] Add a config.base hard-coded to the testing needs somewhere
- [x] Analysis output link not created for some reason ...
~- [ ] fms io, not sure I care at this point~
- [x] copy backgrounds from comrot to rundir 

Not really needed yet:
- [x] minor updates to the g-w to be pushed as a pr <-- Not needed unless you run in containers

### PR DESCRIPTION
**Pull request by guillaumevernieres**
Still a bit of a work todo but good enough for a draft PR.

#### Addition of 3 ctests related to running with the g-w
  - Copy/paste of @CoryMartin-NOAA 's ``setup_expt.py`` test adjusted to `soca`
  - jjob for the Marine DA prep step (create yaml's, stage obs and bkg's, ...)
  - jjob for the Marine DA run step (gridgen, static B init., run var, create iau increment, ...)   

#### Stuff left to do
 - Only tested in a singularity container
 ~- As I type this I'm wondering if the changes I made to the g-w are needed (probably not)~
 - The prep jjob fails when run twice in a row, simple fix but still in the todo list
 - ...

#### Issues addressed
- fixes #232

### CODE REVIEW

File: scripts/exgdas_global_marine_analysis_prep.py
Comment (CoryMartin-NOAA) [inline scripts/exgdas_global_marine_analysis_prep.py:47]: FYI, Rahul will want this set at the jjob_prep.sh step to be consistent with `jobs/rocoto/prep.sh` in g-w

File: scripts/exgdas_global_marine_analysis_prep.py
Comment (CoryMartin-NOAA) [inline scripts/exgdas_global_marine_analysis_prep.py:166]: I'm hoping to phase out `ufsda` in favor of `pygw`. I will not expect that of soca/atm just yet though but plan to try to do so for aero

File: test/soca/gw/jjob_prep.sh
Comment (CoryMartin-NOAA) [inline test/soca/gw/jjob_prep.sh:8]: why is this being done here and not part of the prep job?

**Review by CoryMartin-NOAA — COMMENTED** [review]
a few comments/questions

File: test/soca/gw/jjob_prep.sh
Comment (guillaumevernieres) [reply test/soca/gw/jjob_prep.sh:8]: Are you talking about the background for the previous cycle or the sourcing of the env. vars. @CoryMartin-NOAA ?

File: scripts/exgdas_global_marine_analysis_prep.py
Comment (guillaumevernieres) [reply scripts/exgdas_global_marine_analysis_prep.py:166]: OK, thanks for the info.

File: test/soca/gw/jjob_prep.sh
Comment (CoryMartin-NOAA) [reply test/soca/gw/jjob_prep.sh:8]: the background, shouldn't it be 'available' already

Comment (guillaumevernieres) [reply test/soca/gw/jjob_prep.sh:8]: They would be after 1 cycle. But yes, good point ... It's not part of the prep jjob currently. I don't think I understand quite yet how the first cycle is handled in the g-w. I've been copying background files of the first cycle to a specific location to so far.  
Going to wait and see what @RussTreadon-NOAA is doing :)

Comment (CoryMartin-NOAA) [reply test/soca/gw/jjob_prep.sh:8]: I think you are talking about a 'half cycle' vs the 'full cycle', we shouldn't be testing the 'half cycle' because there is no DA yet. Assume there has already been a forecast, and the files you need are in `$COMROT` when you get to your prep step.

Comment (guillaumevernieres) [reply test/soca/gw/jjob_prep.sh:8]: Ha yes ... I forgot the jargon of the first cycle issue. 
So that's what that step above is doing, populating `$COMROT`, do you mean that step should be done in the previous test? Or hidden in the `CMakeLists.txt`?

Comment (CoryMartin-NOAA) [reply test/soca/gw/jjob_prep.sh:8]: `$COMROT` should be `$GDASAPP_TESTDATA/lowres/` I think? Perhaps it needs slightly modified. Or I would suggest as you said a new test that just stages the ROTDIR.

Comment (guillaumevernieres) [reply test/soca/gw/jjob_prep.sh:8]: it's definitely **not a test** and shouldn't be, it doesn't cover any code, just copies files somewhere. But it probably would make sense to bury the creation of the `$COMROT` in the `CMakeList.txt`. Just following the path of least resistance, it's simpler to do `cp` in `bash` than in `cmake` :)

**Comment by guillaumevernieres** [comment]
Looks like I hard-coded `CONTAINER` in `config.base`, so the new tests are failing and ... the file name is 136 character long, so it's truncated when going through fms io. I'm not sure why that bug didn't show up the last time we ran the ci.

**Comment by guillaumevernieres** [comment]
@RussTreadon-NOAA , @CoryMartin-NOAA , I pushed a few bug fixes related to the marine DA in the global-workflow, see [PR 1172](https://github.com/NOAA-EMC/global-workflow/pull/1172). We'll have to wait until this is merged before running the CI again ...

**Comment by guillaumevernieres** [comment]
@RussTreadon-NOAA , @CoryMartin-NOAA : We'll need to wait for the [g-w pr#1172](https://github.com/NOAA-EMC/global-workflow/pull/1172) to be merged before triggering the CI, but it tested OK in the container and Orion.

**Comment by guillaumevernieres** [comment]
That's funny, I did all the testing on Orion and none on Hera ...
@CoryMartin-NOAA  Any chance you could remove the orion pr 234 in your ci directory? The CI is failing because it's using a g-w develop that is a few commits behind and doesn't include the few bug fixes that were merge yesterday evening.

File: scripts/exgdas_global_marine_analysis_prep.py
Comment (CoryMartin-NOAA) [reply scripts/exgdas_global_marine_analysis_prep.py:47]: This comment still applies :-), set this at the j-job with `$PYTHONPATH`

File: scripts/exgdas_global_marine_analysis_prep.py
Comment (CoryMartin-NOAA) [inline scripts/exgdas_global_marine_analysis_prep.py:127]: good for now, but we will want to transition to `pygw` tools ASAP

File: test/soca/gw/bkg2comrot.sh
Comment (CoryMartin-NOAA) [inline test/soca/gw/bkg2comrot.sh]: Use standard variables in the path here? like ROTDIR, PDY, cyc, etc

File: test/soca/gw/jjob_prep.sh
Comment (CoryMartin-NOAA) [inline test/soca/gw/jjob_prep.sh]: Same comment, don't use hard coded path/dates here

File: scripts/exgdas_global_marine_analysis_prep.py
Comment (guillaumevernieres) [reply scripts/exgdas_global_marine_analysis_prep.py:47]: created this issue: [GDASApp/242](https://github.com/NOAA-EMC/GDASApp/issues/242)

File: test/soca/gw/jjob_prep.sh
Comment (guillaumevernieres) [reply test/soca/gw/jjob_prep.sh:8]: I see I wrote it's definitely not a ctest ... But I was convinced otherwise last week. It was added as a separate ctest. Marking as resolved @CoryMartin-NOAA .

File: scripts/exgdas_global_marine_analysis_prep.py
Comment (guillaumevernieres) [reply scripts/exgdas_global_marine_analysis_prep.py:166]: Resolving that one since you addressed that problem in one of the last few PR's @CoryMartin-NOAA .

File: test/soca/gw/bkg2comrot.sh
Comment (guillaumevernieres) [reply test/soca/gw/bkg2comrot.sh]: Cleaned up the run time variables and now making use of pdy/cyc/...

File: test/soca/gw/jjob_prep.sh
Comment (guillaumevernieres) [reply test/soca/gw/jjob_prep.sh]: done as well. I did also run shellcheck since my bash style was all over the place.

File: test/soca/gw/setup_workflow_exp.sh
Comment (CoryMartin-NOAA) [inline test/soca/gw/setup_workflow_exp.sh:11]: this is fine, but this is not equivalent to what was here before, just a comment...

Comment (guillaumevernieres) [reply test/soca/gw/setup_workflow_exp.sh:11]: the dates were wrong.

**Review by CoryMartin-NOAA — DISMISSED** [review]
Looks good to me

File: test/soca/gw/setup_workflow_exp.sh
Comment (guillaumevernieres) [reply test/soca/gw/setup_workflow_exp.sh:11]: It doesn't look like that `setup_expt.py` does anything with the dates. `PDY` and `cyc` from the environment that we set is what's used.

**Comment by guillaumevernieres** [comment]
We shall see, but I think all our test should be re-runable (it's a word ... ) now.

**Comment by guillaumevernieres** [comment]
Re-requesting your review @CoryMartin-NOAA , I had to add a few changes to make some of the ctest "re-runable".

File: test/soca/gw/setup_workflow_exp.sh
Comment (CoryMartin-NOAA) [reply test/soca/gw/setup_workflow_exp.sh:11]: the dates are just used to set up the rocoto XML (I believe)

**Review by CoryMartin-NOAA — APPROVED** [review]

**Pull request merged**
